websocket.go

```diff
@@ -0,0 +1,359 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/coder/websocket"
+)
+
+type WebSocketProxy struct {
+	upstreamURL string
+	clients     map[*websocket.Conn]map[string]bool // Tracks swap IDs for each client
+	subscribers map[string]map[*websocket.Conn]bool // Tracks clients for each swap ID
+	mu          sync.Mutex
+	upstream    *websocket.Conn
+}
+
+func NewWebSocketProxy(upstreamURL string) *WebSocketProxy {
+	return &WebSocketProxy{
+		upstreamURL: upstreamURL,
+		clients:     make(map[*websocket.Conn]map[string]bool),
+		subscribers: make(map[string]map[*websocket.Conn]bool),
+	}
+}
+
+func (p *WebSocketProxy) connectToUpstream() error {
+	conn, _, err := websocket.Dial(context.Background(), p.upstreamURL, nil)
+	if err != nil {
+		return err
+	}
+	p.upstream = conn
+	return nil
+}
+
+func (p *WebSocketProxy) HandleWebSocket(w http.ResponseWriter, r *http.Request) {
+	conn, err := websocket.Accept(w, r, nil)
+	if err != nil {
+		log.Printf("WebSocket upgrade error: %v", err)
+		return
+	}
+	defer conn.CloseNow()
+
+	// Set a read timeout (e.g., 30 seconds)
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel() // Ensure the context is canceled on all paths
+
+	p.mu.Lock()
+	p.clients[conn] = make(map[string]bool) // Initialize swap IDs for this client
+	p.mu.Unlock()
+
+	defer func() {
+		p.mu.Lock()
+		defer p.mu.Unlock()
+
+		// Remove client from all subscriptions
+		for swapID := range p.clients[conn] {
+			delete(p.subscribers[swapID], conn)
+			if len(p.subscribers[swapID]) == 0 {
+				// If no more clients are subscribed to this swap ID, remove it from the map
+				delete(p.subscribers, swapID)
+
+				// Optionally, send an unsubscribe message to the upstream server
+				if p.upstream != nil {
+					unsubscribeMsg := map[string]any{
```

Comment:
unsubscribe supports multiple swaps like the code does in subscribe. It seems better to aggregate the swap ids and call once at the end of the loop to unsubscribe.